File: src/main/java/com/pandus/leetcode/solutions/google/ThreeSum.java
package com.pandus.leetcode.solutions.google;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        int[] array = Arrays.stream(nums).sorted().toArray();
        Set<List<Integer>> result = new HashSet<>();
        for (int i = 0; i < array.length - 1; i++) {
            int first = array[i];
            int target = -first;
            int left = i + 1;
            int right = array.length - 1;
            while (left < right) {
                int sum = array[left] + array[right];
                if (sum == target) {
                    List<Integer> triplet = new ArrayList<>();
                    triplet.add(first);
                    triplet.add(array[left]);
                    triplet.add(array[right]);
                    result.add(triplet);
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return new ArrayList<>(result);
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/top/interview/ThreeSum.kt
package com.pandus.leetcode.solutions.top.interview

// Reference: https://leetcode.com/problems/3sum/description/
class ThreeSum {
    fun threeSum(nums: IntArray): List<List<Int>> {
        val nums = nums.sorted()
        val resultSet: MutableSet<List<Int>> = HashSet()
        for (i in 0 until nums.size - 2) {
            if (nums[i] > 0) break

            var left = i + 1
            var right = nums.size - 1

            while (left < right) {
                val sum = nums[i] + nums[left] + nums[right]
                when {
                    sum == 0 -> {
                        resultSet.add(listOf(nums[i], nums[left], nums[right]))
                        left++
                        right--
                    }

                    sum < 0 -> left++
                    else -> right--
                }
            }
        }
        return resultSet.map { it.toList() }
    }
}
